### PR TITLE
Resolve dependencies with a Scala 3 partial version

### LINKED ISSUE #2407: `--scala-version 3` returns 2.13 artefacts
Using Coursier 2.1.0-M5-18-gfebf9838c :
```shell
% cs fetch --scala-version 3 'org.scalacheck::scalacheck:1.16.0'
/home/smarter/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scalacheck/scalacheck_2.13/1.16.0/scalacheck_2.13-1.16.0.jar
/home/smarter/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.8/scala-library-2.13.8.jar
/home/smarter/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/test-interface/1.0/test-interface-1.0.jar
```
(Also IMO it's time to change the default to scala 3 from scala 2.13 now that `cs launch scala` starts Scala 3).

### PR DESCRIPTION
**Pull request by julienrf**
Fixes #2407

- Support partial versions such as "3" (previously, you had to write at least two numbers, as in "2.13")
- Fix module name of forced Scala modules (scala3-library and scala3-compiler). This also fixes a bug in "cs fetch --scala-version 3.2.1 '...'" not forcing the Scala 3 version in scala3-library)
- Look up Scala 3 versions (previously, only Scala 2 versions were looked up), and default to Scala 3 if the module has been published for Scala 3.

### CODE REVIEW

File: modules/cli-tests/src/main/scala/coursier/clitests/FetchTests.scala [hunk 1]
```diff
@@ -90,5 +90,32 @@ abstract class FetchTests extends TestSuite {
...
+    }
+
+    test("Scala 2 partial version with one number") {
+      val res0 =
+        os.proc(launcher, "fetch", "org.scalacheck::scalacheck:1.16.0", "--scala-version", "2")
+          .call()
+      assert(res0.exitCode == 0)
+      val output = res0.out.text()
+      assert(output.contains("org/scalacheck/scalacheck_2.13/1.16.0") && output.contains("org/scala-lang/scala-library/2.13."))
+    }
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply modules/cli-tests/src/main/scala/coursier/clitests/FetchTests.scala:119]: I think @alexarchambault is looking into that

**Review by tgodzik — APPROVED** [review]
Super minor from me, but looks good!

File: modules/coursier/shared/src/main/scala/coursier/parse/JavaOrScalaModule.scala [hunk 2]
```diff
@@ -27,7 +27,7 @@ object JavaOrScalaModule {
 
   def scalaBinaryVersion(scalaVersion: String): String =
     // Directly inspired from https://github.com/sbt/librarymanagement/blob/5ef0af2486d19cc237684000ef34c99191b88dfd/core/src/main/scala/sbt/internal/librarymanagement/cross/CrossVersionUtil.scala#L87
-    if (scalaVersion.startsWith("3."))
+    if (scalaVersion.startsWith("3"))
```
[the remark this answers is not in the document]
Comment (julienrf) [reply modules/coursier/shared/src/main/scala/coursier/parse/JavaOrScalaModule.scala]: There will be other things to change anyway to support Scala 4 but I applied your suggestion in de79bba8965a554f85f24bb8b54baf66dc8bf78a

**Comment by julienrf** [comment]
Note that another change in this PR is that now Coursier defaults to Scala 3 if no Scala version is provided:

~~~ text
$ ./mill -i cli.run fetch org.scalacheck::scalacheck:1.16.0
[467/467] cli.run 
/home/julien/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scalacheck/scalacheck_3/1.16.0/scalacheck_3-1.16.0.jar
/home/julien/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.1.1/scala3-library_3-3.1.1.jar
/home/julien/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/test-interface/1.0/test-interface-1.0.jar
/home/julien/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.6/scala-library-2.13.6.jar
~~~

File: modules/cli-tests/src/main/scala/coursier/clitests/FetchTests.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexarchambault) [reply modules/cli-tests/src/main/scala/coursier/clitests/FetchTests.scala:127]: Yeah, that looks like what https://github.com/coursier/coursier/pull/2607 should address.

**Review by alexarchambault — APPROVED** [review]
Merging, thanks @julienrf!

**Pull request merged**